File: TimeDilation/timeDilation.py
#! /usr/bin/env python3

import math
from decimal import *

# https://docs.python.org/3/library/decimal.html

def main():

    getcontext().prec = 32

    mphToMs   = Decimal("0.44704")
    c         = Decimal("299792458")
    c_sqrd    = c ** 2
    
    gamma     = 0.0
    gamma_inv = 0.0

    temp     = 0.0
    proptime = 0.0
    time     = 1.0 #in hours

    diff     = 0.0

    time = Decimal((2 * 5) * (16 * 2) * 3600) #time driving to/from odu in one year

    for speed in range(10, 65, 5):
        speed_in_ms = speed * mphToMs

        gamma     = 1.0 / math.sqrt( 1 - (speed_in_ms / c ** Decimal(2.0)) ) 
        gamma_inv = Decimal( c_sqrd - Decimal(Decimal(0.5) * speed_in_ms * speed_in_ms) ) / c_sqrd
        proptime  = time / gamma_inv

        diff = time - proptime
        
        print("-" * 40)
        print("MPH          ", speed)
        print("m/s          ", speed_in_ms)
        print("Gamma        ", gamma)
        print("Gamma inv    ", gamma_inv)
        print("Time         ", time)
        print("Time(proper) ", proptime)
        print("Difference   ", diff)

if __name__ == "__main__":
    main()

File: TimeDilation/TimeDilation.cpp
#include <cmath>
#include <iostream>
#include <iomanip>
#include <string>

using namespace std;

// Constants
const double MPH_TO_M_DIV_S = 0.44704;
const double C              = 299791958;  // speed of light (m/s)
const double C_SQUARED      = C * C;

const int DAY_PER_WEEK   = 5;       // Based on 5-day work-week
const int WEEKS_PER_YEAR = 16 * 2;  // Based on fall & spring semesters
const int DRIVE_DURATION = 60;      // length of drive one way in minutes
const int MINUTES_TO_SECONDS = 60;

const std::string DIVIDER(32, '-');
//------------------------------------------------------------------------------

/**
 * Compute gamma and inverse gamma
 *
 * @tparam T precision (float or double).
 *
 * @param speed travel speed in MPH
 */
template<typename T>
std::pair<T, T> compute_gamma(const T speed)
{
     const T gamma     = 1.0 / sqrt(1 - pow((speed / C), 2));
     const T gamma_inv = (C_SQUARED - (0.5 * pow(speed, 2.0))) / C_SQUARED;

     return {gamma, gamma_inv};
}

//------------------------------------------------------------------------------
int main(int argc, char**  argv)
{
    cout.setf(ios::fixed);
    setprecision(32);

    constexpr double time = 2 * DAY_PER_WEEK * WEEKS_PER_YEAR
                          * DRIVE_DURATION * MINUTES_TO_SECONDS;

    for (double speed = 10; speed <= 60; speed += 5)
    {
        const double speed_in_ms = speed * MPH_TO_M_DIV_S;
        const auto& [gamma, gamma_inv] = compute_gamma<double>(speed_in_ms);

        const double proper_time = time / gamma_inv;
        const double diff = time - proper_time;

        cout << "Mph          " << right << setw(19) << speed       << "\n";
        cout << "m/s          " << right << setw(19) << speed_in_ms << "\n";

        cout << DIVIDER << "\n";
        cout << "Gamma        " << right << setw(19) << gamma       << "\n";
        cout << "Gamma inv    " << right << setw(19) << gamma_inv   << "\n";

        cout << DIVIDER << "\n";
        cout << "Time         " << right << setw(19) << time        << "\n";
        cout << "Time(proper) " << right << setw(19) << proper_time << "\n";
        cout << "Difference   " << right << setw(19) << diff        << "\n"
             << "\n";
    }

    return 0;
}
